certgen/csr_cryptocb.c

```diff
@@ -0,0 +1,505 @@
+        case WC_PK_TYPE_RSA_KEYGEN: return "RSA KeyGen";
+        case WC_PK_TYPE_EC_KEYGEN: return "ECC KeyGen";
+    }
+    return NULL;
+}
+#endif /* DEBUG_CRYPTOCB */
+
+/* reads file size, allocates buffer, reads into buffer, returns buffer */
+static int load_file(const char* fname, byte** buf, size_t* bufLen)
+{
+    int ret;
+    long int fileSz;
+    XFILE lFile;
+
+    if (fname == NULL || buf == NULL || bufLen == NULL)
+        return BAD_FUNC_ARG;
+
+    /* set defaults */
+    *buf = NULL;
+    *bufLen = 0;
+
+    /* open file (read-only binary) */
+    lFile = XFOPEN(fname, "rb");
+    if (!lFile) {
+        printf("Error loading %s\n", fname);
+        return BAD_PATH_ERROR;
+    }
+
+    fseek(lFile, 0, SEEK_END);
+    fileSz = (int)ftell(lFile);
+    rewind(lFile);
+    if (fileSz  > 0) {
+        *bufLen = (size_t)fileSz;
+        *buf = (byte*)malloc(*bufLen);
+        if (*buf == NULL) {
+            ret = MEMORY_E;
+            printf("Error allocating %lu bytes\n", (unsigned long)*bufLen);
+        }
+        else {
+            size_t readLen = fread(*buf, *bufLen, 1, lFile);
+
+            /* check response code */
+            ret = (readLen > 0) ? 0 : -1;
+        }
+    }
+    else {
+        ret = BUFFER_E;
+    }
+    fclose(lFile);
+
+    return ret;
+}
+
+static int load_key_file(const char* fname, byte* derBuf, word32* derLen,
+    int isPubKey)
+{
+    int ret;
+    byte* buf = NULL;
+    size_t bufLen;
+
+    ret = load_file(fname, &buf, &bufLen);
+    if (ret != 0)
+        return ret;
+
+    if (isPubKey) {
+        ret = wc_PubKeyPemToDer(buf, (word32)bufLen, derBuf, (word32)bufLen);
+    } else {
+        ret = wc_KeyPemToDer(buf, (word32)bufLen, derBuf, (word32)bufLen, NULL);
+    }
+    if (ret < 0) {
+        free(buf);
+        return ret;
+    }
+    *derLen = ret;
+    free(buf);
+
+    return 0;
+}
+
+#ifdef WOLFSSL_DER_TO_PEM
+static int save_der_as_pem(const byte* der, word32 derSz, const char* arg1,
+    int type)
+{
+    int ret;
+    byte pem[LARGE_TEMP_SZ];
+    int  pemSz;
+    FILE* file = NULL;
+    char outFile[255];
+
+    memset(pem, 0, sizeof(pem));
+    ret = wc_DerToPem(der, derSz, pem, sizeof(pem), type);
+    if (ret <= 0) {
+        printf("CSR DER to PEM failed: %d\n", ret);
+        return ret;
+    }
+    pemSz = ret;
+    printf("%s (%d)\n", pem, pemSz);
+
+    snprintf(outFile, sizeof(outFile), "%s-csr.pem", arg1);
+    printf("Saved CSR PEM to \"%s\"\n", outFile);
+    file = fopen(outFile, "wb");
+    if (file) {
+        ret = (int)fwrite(pem, 1, pemSz, file);
+        if (ret == pemSz) {
+            ret = 0;
+        }
+        fclose(file);
+    }
+    else {
+        ret = -1;
+    }
+    return ret;
+}
+#endif
+
+/* Example crypto dev callback function that calls software version */
+/* This is where you would plug-in calls to your own hardware crypto */
+static int myCryptoDevCb(int devIdArg, wc_CryptoInfo* info, void* ctx)
```

Comment:
might I suggest forward declaration of all the functions above here and then having this as the first function that is implemented as its basically the functionality being showcased.